src/main/java/graphql/normalized/ExecutableNormalizedOperationFactory.java

```diff
@@ -292,437 +365,451 @@ private ExecutableNormalizedOperation createExecutableNormalizedOperationImplWit
-        for (CollectedField collectedField : fields) {
-            objectTypes.addAll(collectedField.objectTypes);
-        }
-        Set<GraphQLObjectType> allRelevantObjects = objectTypes.build();
-        Map<GraphQLType, ImmutableList<CollectedField>> groupByAstParent = groupingBy(fields, fieldAndType -> fieldAndType.astTypeCondition);
-        if (groupByAstParent.size() == 1) {
-            return singletonList(new CollectedFieldGroup(ImmutableSet.copyOf(fields), allRelevantObjects));
-        }
-        ImmutableList.Builder<CollectedFieldGroup> result = ImmutableList.builder();
-        for (GraphQLObjectType objectType : allRelevantObjects) {
-            Set<CollectedField> relevantFields = filterSet(fields, field -> field.objectTypes.contains(objectType));
-            result.add(new CollectedFieldGroup(relevantFields, singleton(objectType)));
+            this.fieldCount++;
+            if (this.fieldCount > this.options.getMaxFieldsCount()) {
+                throw new AbortExecutionException("Maximum field count exceeded. " + this.fieldCount + " > " + this.options.getMaxFieldsCount());
+            }
+            Field field;
+            Set<GraphQLObjectType> objectTypes = collectedFieldGroup.objectTypes;
+            field = collectedFieldGroup.fields.iterator().next().field;
+            String fieldName = field.getName();
+            GraphQLFieldDefinition fieldDefinition = Introspection.getFieldDef(graphQLSchema, objectTypes.iterator().next(), fieldName);
+
+            Map<String, Object> argumentValues = ValuesResolver.getArgumentValues(fieldDefinition.getArguments(), field.getArguments(), CoercedVariables.of(this.coercedVariableValues.toMap()), this.options.graphQLContext, this.options.locale);
+            Map<String, NormalizedInputValue> normalizedArgumentValues = null;
+            if (this.normalizedVariableValues != null) {
+                normalizedArgumentValues = ValuesResolver.getNormalizedArgumentValues(fieldDefinition.getArguments(), field.getArguments(), this.normalizedVariableValues);
+            }
+            ImmutableList<String> objectTypeNames = map(objectTypes, GraphQLObjectType::getName);
+            return ExecutableNormalizedField.newNormalizedField()
+                    .alias(field.getAlias())
+                    .resolvedArguments(argumentValues)
+                    .normalizedArguments(normalizedArgumentValues)
+                    .astArguments(field.getArguments())
+                    .objectTypeNames(objectTypeNames)
+                    .fieldName(fieldName)
+                    .level(level)
+                    .parent(parent)
+                    .build();
+        }
+
+        private List<CollectedFieldGroup> groupByCommonParents(Collection<CollectedField> fields) {
+            return groupByCommonParentsNoDeferSupport(fields);
+        }
+
+        private List<CollectedFieldGroup> groupByCommonParentsNoDeferSupport(Collection<CollectedField> fields) {
```

Comment:
Keeping in the "no defer" method in

Balance between wanting to make this file look like master (and reduce risk of bad cherry pick) and then having this odd sounding method talking about a feature in the future

I opted for safety